// mt-changelog/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _latestTag = require('./latest-tag');

var _latestTag2 = _interopRequireDefault(_latestTag);

var _gitLog = require('./git-log');

var _gitLog2 = _interopRequireDefault(_gitLog);

var _output = require('./output');

var _output2 = _interopRequireDefault(_output);

exports['default'] = function (options) {
  return (0, _latestTag2['default'])(options.excludePreReleases).then(function (lastTag) {
    return (0, _gitLog2['default'])(lastTag, options.message);
  }).then(function (log) {
    var formattedLog = options.formatter({
      title: options.title ? options.title : '<title not provided>',
      time: new Date(),
      log: log
    });

    (0, _output2['default'])({
      stdout: options.stdout,
      filepath: options.out,
      formattedLog: formattedLog
    });
  });
};

module.exports = exports['default'];
